src/app/tests/TestDataModelSerialization.cpp

```diff
@@ -0,0 +1,872 @@
+                                   MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::SimpleStruct::kEFieldId), t.e));
+                NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+            }
+
+            err = _this->mWriter.Finalize();
+            NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+            _this->DumpBuf();
+        }
+
+        //
+        // Decode
+        //
+        {
+            clusters::TestCluster::SimpleStruct::Type t;
+
+            _this->SetupReader();
+
+            err = DataModel::Decode(_this->mReader, t);
+            NL_TEST_ASSERT(apSuite, err != CHIP_NO_ERROR);
+        }
+    }
+}
+
+void TestDataModelSerialization::TestDataModelSerialization_InvalidListType(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err;
+    auto * _this = static_cast<TestDataModelSerialization *>(apContext);
+
+    _this->mpSuite = apSuite;
+    _this->SetupBuf();
+
+    //
+    // Encode
+    //
+    {
+        clusters::TestCluster::NestedStructList::Type t;
+        uint32_t intBuf[4] = { 10000, 10001, 10002, 10003 };
+
+        t.e = intBuf;
+
+        // Encode a list of integers for field d instead of a list of structs.
+        {
+            err = EncodeStruct(_this->mWriter, TLV::AnonymousTag,
+                               MakeTagValuePair(TLV::ContextTag(clusters::TestCluster::NestedStructList::kDFieldId), t.e));
+            NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+        }
+
+        err = _this->mWriter.Finalize();
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        _this->DumpBuf();
+    }
+
+    //
+    // Decode
+    //
+    {
+        clusters::TestCluster::DecodableNestedStructList::Type t;
+
+        _this->SetupReader();
+
+        err = DataModel::Decode(_this->mReader, t);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        auto iter     = t.d.begin();
+        bool hadItems = false;
+
+        while (iter.Next())
+        {
+            hadItems = true;
+        }
+
+        NL_TEST_ASSERT(apSuite, iter.GetError() != CHIP_NO_ERROR);
+        NL_TEST_ASSERT(apSuite, !hadItems);
+    }
+}
+
+} // namespace app
+} // namespace chip
+
+namespace {
+
+void InitializeChip(nlTestSuite * apSuite)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::Optional<chip::Transport::PeerAddress> peer(chip::Transport::Type::kUndefined);
+    chip::Transport::FabricTable fabrics;
+
+    err = chip::Platform::MemoryInit();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    chip::gSystemLayer.Init();
+
+    err = chip::gSessionManager.Init(&chip::gSystemLayer, &chip::gTransportManager, &fabrics, &chip::gMessageCounterManager);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = chip::gExchangeManager.Init(&chip::gSessionManager);
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    err = chip::app::InteractionModelEngine::GetInstance()->Init(&chip::gExchangeManager, nullptr);
```

Comment:
Or any of these.  Please take this spurious stuff out.  And the corresponding headers should go too.

Reply:
Copy-paste remnants from a test I had for end-to-end testing of commands...